recipes-core/searcher-ssh-key/files/mode-controller.c

```diff
@@ -0,0 +1,391 @@
+            exit(1);
+        }
+
+        if (execute_transition_script(MODE_MAINTENANCE, MODE_PRODUCTION) != 0) {
+            log_message(LOG_ERR, "Transition failed");
+            write_mode(MODE_MAINTENANCE);
+            exit(1);
+        }
+
+        if (write_mode(MODE_PRODUCTION) != 0) {
+            log_message(LOG_ERR, "Failed to set production mode");
+            exit(1);
+        }
+
+        log_message(LOG_INFO, "Successfully transitioned to production mode");
+        exit(0);
+    } else if (child_pid > 0) {  // Parent process
+        printf("Transition to production mode initiated\n");
+        exit(0);
+    } else {
+        log_message(LOG_ERR, "Failed to fork for transition");
+        return 1;
+    }
+
+    return 0;
+}
+
+static int handle_maintenance_transition(void) {
+    char current_mode[32];
+
+    if (read_current_mode(current_mode, sizeof(current_mode)) != 0) {
+        log_message(LOG_ERR, "Failed to read current mode");
+        return 1;
+    }
+
+    if (strcmp(current_mode, MODE_PRODUCTION) != 0) {
+        log_message(LOG_ERR, "Can only transition to maintenance from production mode");
+        release_lock();
+        return 1;
+    }
+
+    log_message(LOG_INFO, "Initiating transition to maintenance mode");
+
+    // Set transition mode first
+    if (write_mode(MODE_TRANSITION) != 0) {
+        log_message(LOG_ERR, "Failed to set transition mode");
+        release_lock();
+        return 1;
+    }
+
+    // Apply transition rules
+    if (execute_transition_script(MODE_PRODUCTION, MODE_MAINTENANCE) != 0) {
+        log_message(LOG_ERR, "Transition failed");
+        write_mode(MODE_PRODUCTION);
+        release_lock();
+        return 1;
+    }
+
+    // Fork for the delay and final transition
+    pid_t child_pid = fork();
+    if (child_pid == 0) {  // Child process
+        // Detach from parent
+        setsid();
+        
+        // Wait for the timeout
+        sleep(TIMEOUT_SECONDS);
+        
+        // Set final mode
+        if (write_mode(MODE_MAINTENANCE) != 0) {
+            log_message(LOG_ERR, "Failed to set maintenance mode");
+            exit(1);
+        }
+        
+        log_message(LOG_INFO, "Successfully transitioned to maintenance mode");
+        exit(0);
+    } else if (child_pid > 0) {  // Parent process
+        log_message(LOG_INFO, "Transition initiated, system is now in transition mode");
+        release_lock();
+        return 0;
+    } else {
+        log_message(LOG_ERR, "Failed to fork for transition delay");
+        write_mode(MODE_PRODUCTION);  // Revert on failure
+        release_lock();
+        return 1;
+    }
+
+    return 0;
+}
+
+int main(int argc, char *argv[]) {
+    openlog("mode-controller", LOG_PID, LOG_DAEMON);
+
+    if (argc < 2) {
+        fprintf(stderr, "Usage: %s {transition-to-maintenance|to-production|status}\n", argv[0]);
+        return 1;
+    }
+
+    if (strcmp(argv[1], "transition-to-maintenance") == 0) {
```

Comment:
General comment that `strcmp` is a buffer overrun waiting to happen - I'd *strongly* recommend `strnlen_s` + `strncmp` for all the comparisons (or C++ but it's a bit too late for that)